src/main/java/com/hyperlocal/server/ShopController.java

```diff
@@ -35,6 +52,144 @@ public ShopController(PubSubTemplate pubSubTemplate) {
     connection = MySQLConnectionBuilder.createConnectionPool(DATABASE_URL);
   }
 
+  /* To-do server-side checks: 
+  - Access control
+  - Data validation */
+
+  // Fetch all shops by merchantID
+  @GetMapping("/api/merchant/{merchantID}/shops")
+  public CompletableFuture<HashMap<String, Object>> getShopsByMerchantID(@PathVariable Long merchantID) {
+    // Container obj for merchant's shops
+    HashMap<String, Object> shopsMap = new HashMap<String, Object>();
+
+    // Promise: returns merchant's shops
+    CompletableFuture<HashMap<String, Object>> shopsPromise = connection
+        // Get associated shops
+      .sendPreparedStatement(SELECT_SHOPS_BY_MERCHANT_STATEMENT, Arrays.asList(merchantID))
+      .thenApply((QueryResult shopQueryResult) -> {        
+        ResultSet shopRecords = shopQueryResult.getRows();
+        ArrayList<Shop> shopsList = new ArrayList<Shop>();
+        for(RowData shopRecord : shopRecords) shopsList.add(new Shop(shopRecord));
+        shopsMap.put("count", shopsList.size());
+        shopsMap.put("shops", shopsList);
+        return shopsMap;
+        
+        // If something goes wrong:
+      }).exceptionally(ex -> {
+        logger.error("Executed exceptionally: getShopsByMerchantID()", ex);
+        return Helper.generateError(ex.getMessage());
+      });
+
+    return shopsPromise;
+  }
+
+
+  // Fetch catalog, shop & merchant details by shopID.
+  @GetMapping("/api/shop/{shopID}")
+  public CompletableFuture<HashMap<String, Object>> getShopDetails(@PathVariable Long shopID) {
+
+    // Container obj for shop details
+    HashMap<String, Object> shopDetailsMap = new HashMap<String, Object>();
+
+    // Promise: returns shop details
+    CompletableFuture<HashMap<String, Object>> shopDetailsPromise = connection
+        // Get Shop details:
+      .sendPreparedStatement(SELECT_SHOP_STATEMENT, Arrays.asList(shopID))
+      .thenCompose((QueryResult shopQueryResult) -> {
+        ResultSet wrappedShopRecord = shopQueryResult.getRows();
+        if(wrappedShopRecord.size() == 0) throw new NullPointerException("Shop not found."); // No shop with supplied ShopID found
+        RowData shopRecord = wrappedShopRecord.get(0);
+        Shop shop = new Shop(shopRecord);
+        shopDetailsMap.put("shopDetails", shop);
+
+        // Get Merchant Details:
+        return connection.sendPreparedStatement(SELECT_MERCHANT_STATEMENT, Arrays.asList(shop.merchantID));
+      }).thenCompose((QueryResult merchantQueryResult) -> {
+        RowData merchantRecord = merchantQueryResult.getRows().get(0);
```

Comment:
You will see NPE if there is not a single row

Reply:
Can you address this ?

Reply:
This code is not accessing a column - get(0) here fetches the first row in the DB query response.

Reply:
Yes, this means that if there is no row, this will fail, ( can that happen, I guess not since for a shop a merchant has to be present). 

It would be good to structure it like 

if(merchantQueryResult.getRows().size == 0 ) throw InternalServerErrorException("something is wrong, no merchant found for a shop)

Do rest of the things.